backend/tools/keycloak/sync/Factory.cs

```diff
@@ -0,0 +1,386 @@
+using System;
+using System.Collections.Generic;
+using System.IdentityModel.Tokens.Jwt;
+using System.Linq;
+using System.Net;
+using System.Net.Http;
+using System.Text;
+using System.Text.Json;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using Pims.Tools.Keycloak.Sync.Exceptions;
+using Pims.Tools.Keycloak.Sync.Extensions;
+using Pims.Tools.Keycloak.Sync.Models;
+using KModel = Pims.Tools.Keycloak.Sync.Models.Keycloak;
+
+namespace Pims.Tools.Keycloak.Sync
+{
+    /// <summary>
+    /// Factory class, provides a way to sync Keycloak roles, groups and users with PIMS.
+    /// </summary>
+    public class Factory : IFactory
+    {
+        #region Variables
+        private readonly ToolOptions _options;
+        private readonly IOpenIdConnector _auth;
+        private readonly HttpClient _client;
+        private readonly JwtSecurityTokenHandler _tokenHandler;
+        private readonly ILogger _logger;
+        private string _token = null;
+        private string _refreshToken = null;
+        private readonly JsonSerializerOptions _serializeOptions = new JsonSerializerOptions()
+        {
+            PropertyNameCaseInsensitive = true,
+            PropertyNamingPolicy = JsonNamingPolicy.CamelCase
+        };
+        #endregion
+
+        #region Constructors
+        /// <summary>
+        /// Creates a new instance of an Factory class, initializes it with the specified arguments.
+        /// </summary>
+        /// <param name="options"></param>
+        /// <param name="auth"></param>
+        /// <param name="clientFactory"></param>
+        /// <param name="tokenHandler"></param>
+        /// <param name="logger"></param>
+        public Factory(IOptionsMonitor<ToolOptions> options, IOpenIdConnector auth, IHttpClientFactory clientFactory, JwtSecurityTokenHandler tokenHandler, ILogger<Factory> logger)
+        {
+            _options = options.CurrentValue;
+            _auth = auth;
+            _client = clientFactory.CreateClient("Pims.Tools.Keycloak.Sync");
+            _tokenHandler = tokenHandler;
+            _logger = logger;
+        }
+        #endregion
+
+        #region Methods
+        /// <summary>
+        /// Sync roles, groups and users in keycloak and PIMS.
+        /// </summary>
+        /// <returns></returns>
+        public async Task<int> SyncAsync()
```

Comment:
I worry a bit about this all being done in one block with no error handling between sections. Is the intent that if any of the steps fail the whole sync should fail?

Reply:
Any failure will abort, as intended.  But yes I would prefer a more robust solution.  But not today, or for a once run job.